domain/network/state/subnet_test.go

```diff
@@ -0,0 +1,413 @@
+	subnetUUID1, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID1,
+		"10.0.1.0/24",
+		"provider-id-1",
+		"provider-network-id",
+		0,
+		[]string{"az0", "az1"},
+		spUUID.String(),
+		nil,
+	)
+	c.Assert(err, gc.ErrorMatches, "UNIQUE constraint failed: provider_network.provider_network_id")
+}
+
+func (s *stateSuite) TestFailAddTwoSubnetsSameProviderID(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	spUUID, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSpace(ctx.Background(), spUUID, "alpha", "foo", []string{})
+	c.Assert(err, jc.ErrorIsNil)
+
+	subnetUUID0, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID0,
+		"10.0.0.0/24",
+		"provider-id",
+		"provider-network-id-0",
+		0,
+		[]string{"az0", "az1"},
+		spUUID.String(),
+		nil,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	subnetUUID1, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID1,
+		"10.0.1.0/24",
+		"provider-id",
+		"provider-network-id-1",
+		0,
+		[]string{"az0", "az1"},
+		spUUID.String(),
+		nil,
+	)
+	c.Assert(err, gc.ErrorMatches, "UNIQUE constraint failed: provider_subnet.provider_id")
+}
+
+func (s *stateSuite) TestRetrieveSubnetByUUID(c *gc.C) {
+	st := NewState(s.TxnRunnerFactory())
+
+	// Add a subnet of type base.
+	subnetUUID0, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID0,
+		"192.168.0.0/20",
+		"provider-id-0",
+		"provider-network-id-0",
+		0,
+		[]string{"az0", "az1"},
+		"",
+		nil,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	// Add a subnet of type fan.
+	subnetUUID1, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSubnet(
+		ctx.Background(),
+		subnetUUID1,
+		"10.0.0.0/12",
+		"provider-id-1",
+		"provider-network-id-1",
+		0,
+		[]string{"az2", "az3"},
+		"",
+		&network.FanCIDRs{
+			FanLocalUnderlay: "192.168.0.0/20",
+			FanOverlay:       "10.0.0.0/8",
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	// Add a space with subnet base.
+	spUUID, err := utils.NewUUID()
+	c.Assert(err, jc.ErrorIsNil)
+	err = st.AddSpace(ctx.Background(), spUUID, "space0", "provider-space-id", []string{subnetUUID0.String()})
+	c.Assert(err, jc.ErrorIsNil)
+
+	sn0, err := st.GetSubnet(ctx.Background(), subnetUUID0)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(sn0.ID, gc.Equals, network.Id(subnetUUID0.String()))
+	c.Check(sn0.FanInfo, gc.IsNil)
+	c.Check(sn0.CIDR, gc.Equals, "192.168.0.0/20")
+	c.Check(sn0.ProviderId.String(), gc.Equals, "provider-id-0")
+	c.Check(sn0.ProviderNetworkId.String(), gc.Equals, "provider-network-id-0")
+	c.Check(sn0.ProviderSpaceId.String(), gc.Equals, "provider-space-id")
+	c.Check(sn0.SpaceID, gc.Equals, spUUID.String())
+	c.Check(sn0.SpaceName, gc.Equals, "")
+	c.Check(sn0.AvailabilityZones, gc.HasLen, 2)
+	c.Check(sn0.AvailabilityZones, jc.SameContents, []string{"az0", "az1"})
```

Comment:
This seems like a smell, I wonder what you're not checking? I think deep equals enforces a contract.

Reply:
I agree with you it's not very pretty. I have to do it this way because the order is not guaranteed on the mapping function (`ToSubnetInfos()`), specially it doesn't guarantees the order of `AvailabilityZones`. Therefore if I do a deep equality check then this test becomes flaky.